lib/openssl/asn1.rb

```diff
@@ -150,30 +276,217 @@ def each(&blk)
+      def to_der
+        raise ASN1Error, "nil expected" unless @value.nil?
+
+        "\x05\x00".b
+      end
+    end
+
+    class Boolean < Primitive
+      def der_value
+        raise TypeError, "Can't convert nil into Boolean" if @value.nil?
+
+        @value ? "\xff".b : "\x00".b
+      end
+    end
+
+    class Integer < Primitive
+      def der_value
+        ASN1.put_integer(@value)
+      end
     end
 
-    class Boolean < Primitive ; end
-    class Integer < Primitive ; end
-    class Enumerated < Primitive ; end
+    class Enumerated < Primitive
+      def der_value
+        ASN1.put_integer(@value)
+      end
+    end
 
     class BitString < Primitive
+      ASN1_STRING_FLAG_BITS_LEFT = 0x08
+
       attr_accessor :unused_bits
 
       def initialize(*)
         super
 
         @unused_bits = 0
       end
+
+      def der_value
+        if @unused_bits < 0 || @unused_bits > 7
+          raise ASN1Error,  "unused_bits for a bitstring value must be in " \
+		        "the range 0 to 7"
+        end
+
+        return "\x00".b if @value.empty?
+
+        @unused_bits.chr << super
+      end
+    end
+
+    class OctetString < Primitive
+    end
+
+    class UTF8String < Primitive
+    end
+
+    class NumericString < Primitive
+    end
+
+    class PrintableString < Primitive
+    end
+
+    class T61String < Primitive
+    end
+
+    class VideotexString < Primitive
+    end
+
+    class IA5String < Primitive
+    end
+
+    class GraphicString < Primitive
+    end
+
+    class ISO64String < Primitive
+    end
+
+    class GeneralString < Primitive
+    end
+
+    class UniversalString < Primitive
+    end
+
+    class BMPString < Primitive
+    end
+
+    class ObjectId < Primitive
+      def der_value
+        value = oid.split(".").map(&:to_i)
+
+        return (40 * value[0]).chr if value.length == 1
+
+        [value[0] * 40 + value[1], *value[2..]].pack("w*")
+      end
+    end
+
+    class UTCTime < Primitive
+      def der_value
+        value = if @value.is_a?(Time)
+          @value
+        else
+          Time.at(Integer(@value))
+        end
+
+        value.utc.strftime("%y%m%d%H%M%SZ")
+      end
+    end
+
+    class GeneralizedTime < Primitive
+      def der_value
+        value = if @value.is_a?(Time)
+          @value
+        else
+          Time.at(Integer(@value))
+        end
+
+        value.utc.strftime("%Y%m%d%H%M%SZ")
+      end
     end
 
     class EndOfContent < ASN1Data
       def initialize
         super("", 0, :UNIVERSAL)
       end
+
+      def to_der
+        "\x00\x00".b
+      end
+    end
+
+    class Set < Constructive
+    end
+
+    class Sequence < Constructive
+
+    end
+
+    module_function
+
+    # ruby port of openssl ASN1_put_object
+    def put_object(constructed, indefinite_length, length, tag, tag_class)
+      str = "".b
+      xclass = take_asn1_tag_class(tag_class)
+
+      i = constructed ? V_ASN1_CONSTRUCTED : 0
+      i |= (xclass & V_ASN1_PRIVATE)
```

Comment:
`& V_ASN1_PRIVATE` doesn't change the value, as a tag class can only be one of the 4 classes.

Reply:
I ported this from [here](https://github.com/openssl/openssl/blob/454ca902c7d5337249172b38efc5e4fd63f483f4/crypto/asn1/asn1_lib.c#L167)